Allow loaders to define custom script configurations

TheSilkMiner opened the pull request:
Currently, all loaders get access to the same classes and bracket handlers as every other loader. While this might be desirable, it does not make sense for some of them. E.g., the ContentTweaker loader runs too early to have access to recipes, so attempting to access them can lead to issues.

This PR lets loaders dictate what can be loaded and by whom, while also retaining backwards compatibility with the previous behavior. In 1.19, the API will likely get cleaned up.

TODO: ability to mark certain classes as to be registered on certain loaders only through ZenRegister.

TheSilkMiner committed:
Allow different script loaders to have different configs

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

TheSilkMiner committed:
Fix expansions being broken

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/CraftTweakerCommon.java:
```diff
@@ -60,6 +67,16 @@ public static void init() {
         }).start();
     }
     
+    public static void registerCraftTweakerBracketHandlers(final BiConsumer<String, BracketExpressionParser> reg) {
```
Do we really want a consumer here, or another FU (with the same semantics) that describes what the two types stand for?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/ScriptLoadingOptions.java:
```diff
@@ -1,19 +1,28 @@
     private boolean execute;
     private String loaderName = CraftTweakerConstants.DEFAULT_LOADER_NAME;
     private ScriptLoadSource source = RELOAD_LISTENER_SCRIPT_SOURCE;
+    private LegacyRunConfig runConfiguration = this::legacyConfig;
```
We may need to think about the namings:

We **load Scripts** 
via **Script Runs** 
that have **ScriptLoadingOptions**
which rely on a **LoaderName** and a **RunConfig**

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/ScriptLoadingOptions.java:
```diff
@@ -111,8 +120,51 @@ public ScriptLoadingOptions setSource(ScriptLoadSource source) {
+    public interface ScriptRunConfiguration {
+        
+        void configureRun(
+                final BiConsumer<String, BracketExpressionParser> registrationFunction,
+                final Consumer<JavaNativeModule> addingFunction,
```
Names: 
- registrationFunction: register what where?
- addingFunction: adding What to what?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/ScriptLoadingOptions.java:
```diff
@@ -111,8 +120,51 @@ public ScriptLoadingOptions setSource(ScriptLoadSource source) {
-    public static record ScriptLoadSource(ResourceLocation id) {
+    public ScriptLoadingOptions setRunConfiguration(final ScriptRunConfiguration runConfiguration) {
+        
+        this.runConfiguration = LegacyRunConfig.adapt(runConfiguration);
```
Why adapt the new to the old instead of the old to the new?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/ScriptLoadingOptions.java:
```diff
@@ -111,8 +120,51 @@ public ScriptLoadingOptions setSource(ScriptLoadSource source) {
+    }
+    
+    @Deprecated(forRemoval = true)
+    private void legacyConfig(final IgnorePrefixCasingBracketParser bep, final ScriptingEngine engine) throws CompileException {
```
So, we create a run config that whenever called creates another run config and executes it?
If the log message and the old event are the only reason for that, why not create a child class of the default config that does these two before calling `super...`

Edit: But for that you would need to implement the default mechanism as class and not as method reference.

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/CraftTweakerDefaultScriptRunConfiguration.java:
```diff
@@ -0,0 +1,109 @@
+
+public final class CraftTweakerDefaultScriptRunConfiguration {
+    
+    public static final ScriptLoadingOptions.ScriptRunConfiguration DEFAULT_CONFIGURATION = CraftTweakerDefaultScriptRunConfiguration::configure;
```
Why not just **implement** the interface ?!?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/CraftTweakerDefaultScriptRunConfiguration.java:
```diff
@@ -0,0 +1,109 @@
+    }
+    
+    private static JavaNativeModule addModule(
+            final BiConsumer<String, BracketExpressionParser> registrationFunction,
+            final Consumer<JavaNativeModule> addingFunction,
+            final Consumer<JavaNativeModule> applyFunction,
```
NAMES.
Even more so for the types.

I am pretty sure that in a few months we will have no idea whatsoever what the difference between those two consumers would be when only looking at the method signature.

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/loader/ScriptRun.java:
```diff
@@ -101,127 +89,81 @@ public void run() throws Exception {
-        bep.register("recipetype", new RecipeTypeBracketHandler(recipeManagers));
-        bep.register("constant", new EnumConstantBracketHandler());
-        
+        Services.EVENT.fireCustomRegisterBepEvent(this.scriptLoadingOptions.getLoaderName(), this.bep::register);
```
For being future-proof should we just send the whole scriptLoadingOptions object?
Is there anything in it that others may distinct on?
Like, special handling if it formats/executes or depending on the LoaderSource?

Though, on second thought it does sound like YAGNI, what do you think 🤔

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/loader/ScriptRun.java:
```diff
@@ -101,127 +89,81 @@ public void run() throws Exception {
         
         //  toggle this to format scripts. Set by /ct format
-        if(scriptLoadingOptions.isFormat()) {
+        if(this.scriptLoadingOptions.isFormat()) {
```
Ui, someone applied CodeStyle rules 🤔

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/loader/ScriptRun.java:
```diff
@@ -101,127 +89,81 @@ public void run() throws Exception {
             
         } else if(CraftTweakerAPI.DEBUG_MODE) {
-            scriptingEngine.createRunUnit().dump(new File("classes"));
+            this.scriptingEngine.createRunUnit().dump(Paths.get("classes").toFile());
```
What was the difference between new Files directly and Paths.get again?
I remember falling back to `Paths.get` at some times but what was the issue again?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/impl/loader/ScriptRun.java:
```diff
@@ -101,127 +89,81 @@ public void run() throws Exception {
-        FileFormatter formatter = new FileFormatter(builder.build());
-        List<ScriptBlock> blocks = scripts.scripts;
-        for(ScriptBlock block : blocks) {
+        scripts.scripts.forEach(block -> {
```
And... why did you convert that foreach to a forEach?
Or was that applied via codestyle rules?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/impl/script/RecipeManagerScriptLoader.java:
```diff
@@ -29,7 +30,9 @@ public static void loadScriptsFromManager(RecipeManager manager) {
-        final ScriptLoadingOptions scriptLoadingOptions = new ScriptLoadingOptions().setSource(ScriptLoadingOptions.CLIENT_RECIPES_UPDATED_SCRIPT_SOURCE)
+        final ScriptLoadingOptions scriptLoadingOptions = new ScriptLoadingOptions()
+                .setSource(ScriptLoadingOptions.CLIENT_RECIPES_UPDATED_SCRIPT_SOURCE)
+                .setRunConfiguration(CraftTweakerDefaultScriptRunConfiguration.DEFAULT_CONFIGURATION)
```
Do we want this in a `.setDefaultRunConfiguration()` call?
Then we have more freedom moving the other class later, and most modders would probably only ever use that instead of the other one anyways 🤔

kindlich commented on Fabric/src/main/java/com/blamejared/crafttweaker/CraftTweakerFabric.java:
```diff
@@ -38,6 +40,16 @@ public void onInitialize() {
         
         Services.PLATFORM.registerCustomTags();
         
+        CraftTweakerEvents.REGISTER_CUSTOM_BEP_EVENT.register((loader, regFun) -> {
```
So on Fabric there's no `@EventListener` like on Forge?

jaredlll08 replied:
Correct, you also can't unregister events on fabric

kindlich commented on Fabric/src/main/java/com/blamejared/crafttweaker/api/event/CraftTweakerEvents.java:
```diff
@@ -5,19 +5,26 @@
     public static final Event<Consumer<IGatherReplacementExclusionEvent>> GATHER_REPLACEMENT_EXCLUSION_EVENT = EventFactory.createArrayBacked(Consumer.class, listeners -> event -> Arrays.stream(listeners)
             .forEach(handler -> handler.accept(event)));
     
+    public static final Event<BiConsumer<String, BiConsumer<String, BracketExpressionParser>>> REGISTER_CUSTOM_BEP_EVENT =
```
Names and Types!!
Because everyone who reads `Event<BiConsumer<String, BiConsumer<String, BracketExpressionParser>>>` will be like "oh yeah, I know what of the type is supposed to do.


Edit: Why no custom event type here like on Forge?

kindlich commented on Fabric/src/main/java/com/blamejared/crafttweaker/impl/util/WrappingBracketParser.java:
```diff
@@ -0,0 +1,32 @@
+        
+        return Collections.emptySet();
+    }
+    
```
Since from what I infer this BEP is not meant to ever do any actual parsing, should we also override the parse function and throw an Exception there?

kindlich commented on Forge/src/main/java/com/blamejared/crafttweaker/CraftTweakerForge.java:
```diff
@@ -15,11 +20,27 @@ public CraftTweakerForge() {
+        
+        if(CraftTweakerConstants.MOD_ID.equals(event.getCurrentLoader())) {
+            // Fire the old event for compatibility: this is only required for the CrT loader
+            Services.EVENT.fireRegisterBEPEvent(new WrappingBracketParser(event::registerBep));
```
What will happen here for a mod that implements both event listeners and adds a BEP in each of them with the same name?
Unlikely (since few mods even have custom event listeners) but just so that we talked about it ^^

kindlich commented on Fabric/src/main/java/com/blamejared/crafttweaker/impl/util/WrappingBracketParser.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+public final class WrappingBracketParser extends IgnorePrefixCasingBracketParser {
```
Is this class something that can be moved to common, since the class exists on both Fabric and Forge side?

kindlich commented on Forge/src/main/java/com/blamejared/crafttweaker/impl/util/WrappingBracketParser.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+public final class WrappingBracketParser extends IgnorePrefixCasingBracketParser {
```
Is this class something that can be moved to common, since the class exists on both Fabric and Forge side?

kindlich commented on Fabric/src/main/java/com/blamejared/crafttweaker/CraftTweakerFabric.java:
```diff
@@ -38,6 +40,16 @@ public void onInitialize() {
+            if(CraftTweakerConstants.MOD_ID.equals(loader)) {
+                // Fire the old event for compatibility: this is only required for the CrT loader
+                Services.EVENT.fireRegisterBEPEvent(new WrappingBracketParser(regFun));
+            }
```
When reading I was confused about where the BEPs would be added here, until I figured out that they are implicitly added by the old event listener, maybe add a comment here to state that 😉

TheSilkMiner committed:
Rework system to add a concept of inheritance

This lets people get the CraftTweaker default BEPs and similar without
requiring them to refire the event for CraftTweaker. This also lets
loaders inherit from other ones as a side effect.

Signed-off-by: TheSilkMiner <thesilkminer@outlook.com>

jaredlll08 commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/scriptrun/IScriptLoadingOptionsView.java:
```diff
@@ -0,0 +1,30 @@
+
+public interface IScriptLoadingOptionsView {
+    
+    boolean isFormat();
+    
+    boolean isExecute();
```
Thoughts on renaming to:
```
shouldFormat()

shouldExecute()
```?

TheSilkMiner replied:
That would break API, unless I were to add them as separate methods that reference the old ones

jaredlll08 replied:
Ahh okay, this is fine then

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/ScriptLoadingOptions.java:
```diff
@@ -111,8 +132,64 @@ public ScriptLoadingOptions setSource(ScriptLoadSource source) {
+    }
+    
+    @Deprecated(forRemoval = true) // TODO("Remove in 1.19")
+    public ScriptLoadingOptions noInherit() {
+        
+        this.inheritedLoaders.clear();
+        return this;
+    }
+    
+    public ScriptLoadingOptions inheritFromLoader(final String... otherLoaders) {
+        
+        this.inheritedLoaders.addAll(Arrays.asList(otherLoaders));
+        return this;
+    }
```
Why two methods instead of setting the collection directly?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/scriptrun/IBepRegistrationHandler.java:
```diff
@@ -0,0 +1,17 @@
+    void registerBracketHandler(final String bepName, final BracketExpressionParser parser);
+    
+    // TODO("Public API?")
+    default void registerBracketHandler(final ValidatedEscapableBracketParser parser) {
```
Who calls this method and is it something that belongs here, or something where the caller could directly call the other method?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/scriptrun/ICustomBepRegistrationEvent.java:
```diff
@@ -0,0 +1,36 @@
+    }
+    
+    // TODO("Public API?")
+    default void registerBepIfLoader(final String loader, final ValidatedEscapableBracketParser parser) {
```
I doubt people will add their own `ValidatedEscapableBracketParser`s 🤔

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/scriptrun/IScriptLoadingOptionsView.java:
```diff
@@ -0,0 +1,30 @@
+    
+    default Collection<String> getTargetedLoaders() {
+        
+        return Stream.concat(this.getInheritedLoaders().stream(), Stream.of(this.getLoaderName())).toList();
```
Not that it makes a difference (since we return Collection), but should our loader be first or last in the returned list.
But since we return a list we have no order guarantees anyways.

Just in case `.distinct()` or `.toSet()`?

kindlich commented on Common/src/main/java/com/blamejared/crafttweaker/api/zencode/scriptrun/IScriptLoadingOptionsView.java:
```diff
@@ -0,0 +1,30 @@
+        return Stream.concat(this.getInheritedLoaders().stream(), Stream.of(this.getLoaderName())).toList();
+    }
+    
+    default boolean appliesLoader(final String loader) {
```
`appliesForLoader` or `appliesToLoader` ?

Or to these options apply the loader (to something)?

kindlich commented on Forge/src/main/java/com/blamejared/crafttweaker/api/event/type/CTCustomBepRegistrationEvent.java:
```diff
@@ -0,0 +1,43 @@
+/**
+ * Allows users to register custom implementations of {@link BracketExpressionParser}.
+ *
+ * <p>This event is invoked every time a new script run is started, even for syntax checks.</p>
```
Speaking of, do we want that in the long run, or would we only do that for first-runs alter and cache all classes/beps/... that belong to a module later?

TheSilkMiner commented:
Trying a new approach that will lead to less of a mess API wise and also not break features like /ct syntax
